release/k3s/k3s.go

```diff
@@ -0,0 +1,568 @@
+	NewK8SVersion string `json:"new_k8s_version"`
+	OldK8SClient  string `json:"old_k8s_client"`
+	NewK8SClient  string `json:"new_k8s_client"`
+	OldK3SVersion string `json:"old_k3s_version"`
+	NewK3SVersion string `json:"new_k3s_version"`
+	ReleaseBranch string `json:"release_branch"`
+	Workspace     string `json:"workspace"`
+	Handler       string `json:"handler"`
+	Email         string `json:"email"`
+	Token         string `json:"token"`
+}
+
+func NewReleaseFromConfig(c *cli.Context) (*Release, error) {
+	configPath := c.String("config")
+	var release *Release
+	if configPath == "" {
+		return nil, errors.New("error: config file required")
+	}
+
+	configData, err := os.ReadFile(configPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(configData, &release); err != nil {
+		return nil, err
+	}
+	return release, nil
+}
+
+// SetupK8sRemotes will clone the kubernetes upstream repo and proceed with setting up remotes
+// for rancher and user's forks, then it will fetch branches and tags for all remotes
+func (r *Release) SetupK8sRemotes(ctx context.Context, ghClient *github.Client) error {
+	k3sDir := filepath.Join(r.Workspace, "kubernetes")
+	_, err := os.Stat(r.Workspace)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(r.Workspace, 0755); err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+	// clone the repo
+	repo, err := git.PlainClone(k3sDir, false, &git.CloneOptions{
+		URL:             k8sUpstreamURL,
+		Progress:        os.Stdout,
+		InsecureSkipTLS: true,
+	})
+
+	if err != nil {
+		if err == git.ErrRepositoryAlreadyExists {
+			logrus.Warnf("Repository already exists")
+			repo, err = git.PlainOpen(k3sDir)
+			if err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+	gitAuth, err := getAuth("")
+	if err != nil {
+		return err
+	}
+	if err := repo.Fetch(&git.FetchOptions{
+		RemoteName:      "origin",
+		Progress:        os.Stdout,
+		Tags:            git.AllTags,
+		InsecureSkipTLS: true,
+	}); err != nil {
+		if err != git.NoErrAlreadyUpToDate {
+			return err
+		}
+	}
+
+	_, err = repo.CreateRemote(&config.RemoteConfig{
+		Name: rancherRemote,
+		URLs: []string{k8sRancherURL},
+	})
+	if err != nil {
+		if err != git.ErrRemoteExists {
+			return err
+		}
+	}
+	logrus.Infof("Remote %s created for url %s, fetching tags", rancherRemote, k8sRancherURL)
+	if err := repo.Fetch(&git.FetchOptions{
+		RemoteName: rancherRemote,
+		Progress:   os.Stdout,
+		Tags:       git.AllTags,
+		Auth:       gitAuth,
+	}); err != nil {
+		if err != git.NoErrAlreadyUpToDate {
+			return err
+		}
+	}
+	userRemoteURL := strings.Replace(k8sUserURL, "user", r.Handler, -1)
+	_, err = repo.CreateRemote(&config.RemoteConfig{
+		Name: r.Handler,
+		URLs: []string{userRemoteURL},
+	})
+	if err != nil {
+		if err != git.ErrRemoteExists {
+			return err
+		}
+	}
+	logrus.Infof("Remote %s created for url %s, fetching tags", r.Handler, userRemoteURL)
```

Comment:
We should limit log output to the upper level caller as this will be weird when called by something other than a CLI. In OpDom we'd have a mix of regular logging and stdout logging which we want to avoid.